### PR TITLE
Fix memory statistics of `MemoryTracker` when input stream released.

### LINKED ISSUE #6130: Metric 'Cop task memory usage(MemoryTracker)' is larger than actual memory usage when storage read thread enabled
a4e5d247-a5c9-4f9e-be9b-7466c8d90c15

7e378822-1cbc-4ae2-b554-505dd959d623

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: close #6130 

Problem Summary:
It is caused by not setting the corresponding memory tracker when the input stream is released (Currently, input streams are managed by `std::shared_ptr`).

### What is changed and how it works?

Explicitly releasing input stream when it is unused.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
 - Run `select count(ol_o_id), count(ol_d_id), count(ol_w_id), count(ol_number), count(ol_i_id), count(ol_supply_w_id), count(ol_delivery_d), count(ol_quantity), count(ol_amount), count(ol_dist_info) from chbenchmark.order_line`
- Before this PR
  - The actual memory usage is about 6.55 - 4.07 = 2.48GB.
  - The peak memory usage of memory tracker is about 60GB.
image
image

- After this PR
  - The actual memory usage is about 4.0 - 1.9 = 2.1GB.
  - The peak memory usage of memory tracker is about 2 GB.
image
image

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by flowbehappy** [comment]
Is the "Cop task memory usage" matched the real memory usage? Better also put the memory metric here to confirm.

File: dbms/src/Storages/DeltaMerge/SegmentReadTaskPool.h [hunk 1]
```diff
@@ -219,6 +219,11 @@ class SegmentReadTaskPool : private boost::noncopyable
         return add_to_scheduler;
     }
 
+    MemoryTrackerPtr & getMemoryTracker()
+    {
+        return mem_tracker;
```
Comment (bestwoody) [inline dbms/src/Storages/DeltaMerge/SegmentReadTaskPool.h:224]: how about add nullptr check? something like: mem_tracker? mem_tracker.get(): nullptr

**Review by bestwoody — APPROVED** [review]
lgtm

**Review by Lloyd-Pottiger — APPROVED** [review]

File: dbms/src/Storages/DeltaMerge/ReadThread/MergedTask.h [hunk 2]
```diff
@@ -116,13 +118,25 @@ class MergedTask
...
+                stream = nullptr;
+            }
+            pool = nullptr;
             finished_count++;
         }
     }
 
+    void setAllStreamsFinished()
+    {
+        for (size_t i = 0; i < units.size(); i++)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/ReadThread/MergedTask.h:135]:
```suggestion
        for (size_t i = 0; i < units.size(); ++i)
```

**Review by flowbehappy — APPROVED** [review]

**Pull request merged**